dev/Kozani/KozaniManager/ConnectionManager.cpp

```diff
@@ -0,0 +1,534 @@
+// Copyright (c) Microsoft Corporation and Contributors.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "pch.h"
+#include <KozaniManager_h.h>
+#include "ConnectionManager.h"
+
+namespace DvcProtocol = Microsoft::Kozani::DvcProtocol;
+
+namespace Microsoft::Kozani::Manager
+{
+    void CALLBACK ConnectionManager::ProcessTerminationCallback(_In_ PVOID parameter, _In_ BOOLEAN timerOrWaitFired) noexcept try
+    {
+        if (timerOrWaitFired)
+        {
+            // Timer timed out. This shouldn't happen as we wait infinite time. 
+            LOG_HR_MSG(E_UNEXPECTED, "Wait timed out tracking process lifetime.");
+            return;
+        }
+
+        ActivationRequestInfo* requestInfo{ reinterpret_cast<ActivationRequestInfo*>(parameter) };
+        ConnectionManager* connectionManager{ requestInfo->connectionManager };
+        auto lock{ connectionManager->m_requestsLock.lock_exclusive() };
+
+        if (connectionManager->m_closing || requestInfo->processLifetimeTrackerDisabled)
+        {
+            // We are closing down or the tracker is disabled. No more processing should be done.
+            return;
+        }
+
+        // Unregister the process lifetime tracker wait handle in a non-blocking way as we no longer need to track the process lifetime.
+        requestInfo->processLifetimeTrackerHandle.reset();
+
+        if (requestInfo->statusCallback)
+        {
+            (void)requestInfo->statusCallback->OnClosed();
+            requestInfo->statusCallback.reset();
+        }
+
+        switch (requestInfo->status)
+        {
+            case RequestStatus::Connecting:
+                for (auto it = connectionManager->m_requestsPendingConnection.begin(); it != connectionManager->m_requestsPendingConnection.end(); it++)
+                {
+                    if (requestInfo == *it)
+                    {
+                        connectionManager->m_requestsPendingConnection.erase(it);
+                        break;
+                    }
+                }
+                break;
+
+            case RequestStatus::Connected:
+            case RequestStatus::AppActivationRequestMade:
+            case RequestStatus::AppActivationSucceeded:
+                connectionManager->SendAppTerminationNotice(requestInfo);
```

Comment:
Yes, it will send the message within the lock. Message sending (requestInfo->dvcChannel->Write()) should be synchronized. Concurrent write is not supported.

Reply:
OK for today but will become a scalability issue in the future as load grows. Please add a comment in the ConnectionManager class' doc-comment (or something equally central) about message request/result is a synchronous turn and must be done while holding the lock, and file a deliverable "Kozani: Support asynchronous/concurrent messages for scalability" for a future day to optimize

Reply:
The limitation is in DVC itself, which is a file handle. You cannot WriteFile from multiple threads without synchronization and expect the data to be coherent. To support concurrent writes, you need multiple DVC channels, like multiple files. You can write to multiple files concurrently but writing to each file itself must be synchronized.